crypto/cloudkms/cloudkms.go

```diff
@@ -0,0 +1,251 @@
+package cloudkms
+
+import (
+	"context"
+	"encoding/asn1"
+	"fmt"
+	"math/big"
+
+	kms "cloud.google.com/go/kms/apiv1"
+	kmspb "google.golang.org/genproto/googleapis/cloud/kms/v1"
+
+	"github.com/onflow/flow-go-sdk"
+	"github.com/onflow/flow-go-sdk/crypto"
+)
+
+// Key is a reference to a Google Cloud KMS asymmetric signing key version.
+//
+// Ref: https://cloud.google.com/kms/docs/creating-asymmetric-keys#create_an_asymmetric_signing_key
+type Key struct {
+	ProjectID  string
+	LocationID string
+	KeyRingID  string
+	KeyID      string
+	KeyVersion string
+}
+
+// ResourceID returns the resource ID for this KMS key version.
+//
+// Ref: https://cloud.google.com/kms/docs/getting-resource-ids
+func (k Key) ResourceID() string {
+	return fmt.Sprintf(
+		"projects/%s/locations/%s/keyRings/%s/cryptoKeys/%s/cryptoKeyVersions/%s",
+		k.ProjectID,
+		k.LocationID,
+		k.KeyRingID,
+		k.KeyID,
+		k.KeyVersion,
+	)
+}
+
+// Client is a client for interacting with the Google Cloud KMS API
+// using types native to the Flow Go SDK.
+type Client struct {
+	client *kms.KeyManagementClient
+}
+
+// NewClient creates a new KMS client.
+func NewClient(ctx context.Context) (*Client, error) {
+	client, err := kms.NewKeyManagementClient(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("cloudkms: failed to initialize client: %w", err)
+	}
+
+	return &Client{
+		client: client,
+	}, nil
+}
+
+// GetPublicKey fetches the public key portion of a KMS asymmetric signing key version.
+//
+// ECDSA_P256 is currently the only Flow signature algorithm supported by Google Cloud KMS.
+//
+// Ref: https://cloud.google.com/kms/docs/retrieve-public-key
+func (c *Client) GetPublicKey(ctx context.Context, key Key) (crypto.PublicKey, crypto.HashAlgorithm, error) {
+	request := &kmspb.GetPublicKeyRequest{
+		Name: key.ResourceID(),
+	}
+
+	result, err := c.client.GetPublicKey(ctx, request)
+	if err != nil {
+		return crypto.PublicKey{},
+			crypto.UnknownHashAlgorithm,
+			fmt.Errorf("cloudkms: failed to fetch public key from KMS API: %v", err)
+	}
+
+	sigAlgo := parseSignatureAlgorithm(result.Algorithm)
+	if sigAlgo == crypto.UnknownSignatureAlgorithm {
+		return crypto.PublicKey{},
+			crypto.UnknownHashAlgorithm,
+			fmt.Errorf(
+				"cloudkms: unsupported signature algorithm %s",
+				result.Algorithm.String(),
+			)
+	}
+
+	hashAlgo := parseHashAlgorithm(result.Algorithm)
+	if hashAlgo == crypto.UnknownHashAlgorithm {
+		return crypto.PublicKey{},
+			crypto.UnknownHashAlgorithm,
+			fmt.Errorf(
+				"cloudkms: unsupported hash algorithm %s",
+				result.Algorithm.String(),
+			)
+	}
+
+	publicKey, err := crypto.DecodePublicKeyPEM(sigAlgo, result.Pem)
+	if err != nil {
+		return crypto.PublicKey{},
+			crypto.UnknownHashAlgorithm,
+			fmt.Errorf("cryptokms: failed to parse PEM public key: %w", err)
+	}
+
+	return publicKey, hashAlgo, nil
+}
+
+// Signer is a Google Cloud KMS implementation of crypto.Signer.
+type Signer struct {
```

Comment:
This file is still pretty compact, so this could be a bit premature, but could consider splitting out the file based on the structs.